### PR TITLE
FIX Don't apply new adapter's inference_mode to active adapters in inject_adapter

### LINKED ISSUE #3487: Adapter injection applies a new adapter's inference mode to active adapters
### System Info

- PEFT `main` at `a5526d27a9d47d1e8264d5e1b1f96c0fdc79464e`
- PEFT 0.20.0
- Python 3.13.13
- PyTorch 2.13.0+cpu
- Transformers 5.14.1
- Windows 11

### Who can help?

@BenjaminBossan

### Reproduction

Four existing strict-xfail tests still reproduce the same adapter-specific trainability problem on current `main`:

```bash
pytest \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_set_correctly_switch_inference_mode \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_load_adapter_then_add_adapter \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_different_targets \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_same_targets \
  --runxfail -q
```

Result: 4 failed.

The failures are:

- Loading a frozen `default` adapter and then loading trainable `other` leaves the `default` parameters trainable instead of `other`.
- Adding another adapter to a model loaded with `is_trainable=False` re-enables gradients on the frozen adapter.
- Adding an AdaLoRA adapter with `inference_mode=True` freezes the still-active trainable adapter, for both same-target and different-target adapters.

These cases appear to share the housekeeping call in `BaseTuner.inject_adapter`:

```python
self.set_adapter(self.active_adapters, inference_mode=peft_config.inference_mode)
```

Here `peft_config` belongs to the adapter being injected, while `active_adapters` can refer to a different adapter. The new adapter's `inference_mode` is therefore applied across adapter boundaries.

If this scope is welcome, I can prepare one focused fix covering the load/add and AdaLoRA cases and turn the four existing xfails into regular regression tests. I would not add or change public API.

### Expected behavior

Injecting or loading an adapter should not change the trainability state of a different adapter. The active adapter should keep its current `requires_grad` state, while the newly loaded adapter should follow its own `is_trainable` or `inference_mode` setting.

### PR DESCRIPTION
**Pull request by Mohith26**
Fixes #3487, reported by @DaoyuanLi2816.

`BaseTuner.inject_adapter`'s housekeeping step called `self.set_adapter(self.active_adapters, inference_mode=peft_config.inference_mode)`, where `peft_config` belongs to the adapter being injected but `active_adapters` usually names different adapters. Adding a frozen adapter therefore froze the active trainable one, and adding a trainable adapter un-froze a deliberately frozen one.

The fix derives the `inference_mode` passed to `set_adapter` from the active adapters' own configs instead of the new adapter's config, keeping each adapter's trainability scoped to its own settings. A freeze-all-then-re-enable approach was rejected because tuners like UniLoRA keep trainable params outside `adapter_layer_names` and only handle them in their `set_adapter` override; the committed single-call approach avoids that class of gaps.

Three of the four strict-xfail tests tracking this now pass and were converted to regular tests, plus a new two-adapter differing-inference_mode regression test. The fourth xfail remains: it additionally expects a non-active adapter loaded with `is_trainable=True` to become trainable, a separate semantic that contradicts currently-passing tests. Relevant classes: 343 passed with a failure list byte-identical to the pre-existing (AdaMSS-related, unrelated) baseline. `make style` clean.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
In the linked issue, the reporter suggested to provide the fix. As such, they have preference. Please don't open PRs proactively if it's not assigned to you.

**Comment by Mohith26** [comment]
You're right, apologies - I missed that the reporter had offered and should have checked. Closing in their favor; if they'd like to reuse anything from this branch they're welcome to it.
